File: go/internal/service/tenant_service.go
package service

import (
	"context"
	"fmt"

	"github.com/hemanthpathath/flex-db/go/internal/repository"
)

// TenantService handles tenant business logic
type TenantService struct {
	repo repository.TenantRepository
}

// NewTenantService creates a new TenantService
func NewTenantService(repo repository.TenantRepository) *TenantService {
	return &TenantService{repo: repo}
}

// Create creates a new tenant
func (s *TenantService) Create(ctx context.Context, slug, name string) (*repository.Tenant, error) {
	if slug == "" {
		return nil, fmt.Errorf("slug is required")
	}
	if name == "" {
		return nil, fmt.Errorf("name is required")
	}

	tenant := &repository.Tenant{
		Slug: slug,
		Name: name,
	}

	return s.repo.Create(ctx, tenant)
}

// GetByID retrieves a tenant by ID
func (s *TenantService) GetByID(ctx context.Context, id string) (*repository.Tenant, error) {
	if id == "" {
		return nil, fmt.Errorf("id is required")
	}
	return s.repo.GetByID(ctx, id)
}

// Update updates an existing tenant
func (s *TenantService) Update(ctx context.Context, id, slug, name, status string) (*repository.Tenant, error) {
	if id == "" {
		return nil, fmt.Errorf("id is required")
	}

	tenant, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}

	if slug != "" {
		tenant.Slug = slug
	}
	if name != "" {
		tenant.Name = name
	}
	if status != "" {
		tenant.Status = status
	}

	return s.repo.Update(ctx, tenant)
}

// Delete deletes a tenant
func (s *TenantService) Delete(ctx context.Context, id string) error {
	if id == "" {
		return fmt.Errorf("id is required")
	}
	return s.repo.Delete(ctx, id)
}

// List retrieves tenants with pagination
func (s *TenantService) List(ctx context.Context, pageSize int32, pageToken string) ([]*repository.Tenant, *repository.ListResult, error) {
	opts := repository.ListOptions{
		PageSize:  int(pageSize),
		PageToken: pageToken,
	}
	return s.repo.List(ctx, opts)
}

File: app/service/tenant_service.py
"""
Tenant service implementation.
"""

from typing import List, Tuple, Optional

from app.repository import Tenant, TenantRepository, ListOptions, ListResult
from app.db.tenant_db_manager import TenantDatabaseManager


class TenantService:
    """Tenant business logic service."""

    def __init__(self, repo: TenantRepository, tenant_db_manager: Optional[TenantDatabaseManager] = None):
        self.repo = repo
        self.tenant_db_manager = tenant_db_manager

    async def create(self, slug: str, name: str) -> Tenant:
        """Create a new tenant and its associated tenant database."""
        if not slug:
            raise ValueError("slug is required")
        if not name:
            raise ValueError("name is required")

        # Create tenant record in control database
        tenant = Tenant(slug=slug, name=name)
        tenant = await self.repo.create(tenant)

        # Create tenant database and run migrations
        if self.tenant_db_manager:
            await self.tenant_db_manager.create_tenant_database(
                tenant_id=tenant.id,
                slug=tenant.slug
            )

        return tenant

    async def get_by_id(self, id: str) -> Tenant:
        """Retrieve a tenant by ID."""
        if not id:
            raise ValueError("id is required")
        return await self.repo.get_by_id(id)

    async def update(self, id: str, slug: str, name: str, status: str) -> Tenant:
        """Update an existing tenant."""
        if not id:
            raise ValueError("id is required")

        tenant = await self.repo.get_by_id(id)

        if slug:
            tenant.slug = slug
        if name:
            tenant.name = name
        if status:
            tenant.status = status

        return await self.repo.update(tenant)

    async def delete(self, id: str) -> None:
        """Delete a tenant."""
        if not id:
            raise ValueError("id is required")
        await self.repo.delete(id)

    async def list(self, page_size: int, page_token: str) -> Tuple[List[Tenant], ListResult]:
        """Retrieve tenants with pagination."""
        opts = ListOptions(page_size=page_size, page_token=page_token)
        return await self.repo.list(opts)
